File: Planning/src/edu/uno/ai/planning/logic/NegatedLiteral.java
package edu.uno.ai.planning.logic;

import edu.uno.ai.planning.MutableState;
import edu.uno.ai.planning.State;

/**
 * The negation of a literal is a literal.
 * 
 * @author Stephen G. Ware
 */
public class NegatedLiteral extends Negation implements Literal {

	/** The literal being negated */
	public final Literal argument;
	
	/**
	 * Constructs a new negated literal.
	 * 
	 * @param argument the literal to be negated
	 */
	public NegatedLiteral(Literal argument) {
		super(argument);
		this.argument = argument;
	}

	@Override
	public Literal substitute(Substitution substitution) {
		return new NegatedLiteral(argument.substitute(substitution));
	}
	
	@Override
	public boolean isTestable() {
		return true;
	}
	
	@Override
	public boolean isTrue(State state) {
		return state.isTrue(this);
	}

	@Override
	public boolean isImposable() {
		return true;
	}

	@Override
	public void impose(MutableState state) {
		state.impose(this);
	}
	
	@Override
	public Literal negate() {
		return argument;
	}
	
	@Override
	public Expression toCNF() {
		return this;
	}

	@Override
	public Expression toDNF() {
		return this;
	}
	
	@Override
	public Expression simplify() {
		return this;
	}
}

File: UnitySokoban/Assets/Scripts/Planning/Planning/Logic/NegatedLiteral.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Planning.Logic
{
    /**
  * The negation of a literal is a literal.
  * 
  * @author Stephen G. Ware
  * @ported Edward Thomas Garcia
  */
    public class NegatedLiteral : Negation, Literal
    {
        /** The literal being negated */
        //public readonly Literal argument;

        /**
         * Constructs a new negated literal.
         * 
         * @param argument the literal to be negated
         */
        public NegatedLiteral(Literal argument) :
           base(argument)
        {
        }

        public override Expression Substitute(Substitution substitution)
        {
            return new NegatedLiteral(argument.Substitute(substitution) as Literal);
        }

        public override bool IsTestable()
        {
            return true;
        }

        public override bool IsTrue(State state)
        {
            return state.isTrue(this);
        }

        public override bool IsImposable()
        {
            return true;
        }

        public override void Impose(MutableState state)
        {
            state.impose(this);
        }

        public override Expression Negate()
        {
            return argument;
        }

        public override Expression ToCNF()
        {
            return NormalForms.toCNF(this);
        }

        public override Expression ToDNF()
        {
            return NormalForms.toDNF(this);
        }

        public override Expression Simplify()
        {
            return this;
        }

        Literal Literal.Substitute(Substitution substitution)
        {
            return new NegatedLiteral(argument.Substitute(substitution) as Literal);
        }

        Literal Literal.Negate()
        {
            return argument as Literal;
        }
    }
}
